### PR TITLE
Magnus/libmpg123 song player

### PR DESCRIPTION
**Pull request by mglyngberg**

### CODE REVIEW

**Review by Ben-Graham — DISMISSED** [review]
Change SongPlayer::getColor to implement multi threading like on the https://github.com/UCSolarCarTeam/Epsilon-Onboard-Media-Control/tree/Ben/MultiThreadColorChange branch. It uses the QConcurrent library.

File: src/BusinessLayer/DeltaSongPlayer.cpp [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+ *  General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General
+ *  Public License along with the Schulich Delta OnBoard Media
+ *  Control.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  For further contact, email <software@calgarysolarcar.ca>
+ */
+
+#include "DeltaSongPlayer.h"
```
Comment (bill-luu) [inline src/BusinessLayer/DeltaSongPlayer.cpp]: Lets rename this guy!

File: src/BusinessLayer/DeltaSongPlayer.cpp [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    buffer = (unsigned char*) malloc(buffer_size * sizeof(unsigned char));
+
+    // set the output format and open the output device
+    format.bits = mpg123_encsize(encoding) * BITS;
+    format.rate = rate;
+    format.channels = channels;
+    format.byte_format = AO_FMT_NATIVE;
+    format.matrix = 0;
+
+    // most important thing used in thread later
```
Comment (bill-luu) [inline src/BusinessLayer/DeltaSongPlayer.cpp]: This comment isn't really that helpful lol, let's explain what this does.

File: src/BusinessLayer/DeltaSongPlayer.cpp [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    return title;
+}
+
+
+double DeltaSongPlayer::getCurrentTime()
+{
+    if (loaded)
+    {
+        off_t length;
+        double times;
```
Comment (bill-luu) [inline src/BusinessLayer/DeltaSongPlayer.cpp]: Let's expand this name to `timeSecs`, similar for minutes.

File: src/BusinessLayer/DeltaSongPlayer.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        return 1;
+    }
+}
+double DeltaSongPlayer::getSongLength()
+{
+    if (loaded)
+    {
+        off_t length;
+        double times;
```
Comment (bill-luu) [inline src/BusinessLayer/DeltaSongPlayer.cpp]: As above.

File: src/BusinessLayer/DeltaSongPlayer.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ *  General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General
+ *  Public License along with the Schulich Delta OnBoard Media
+ *  Control.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  For further contact, email <software@calgarysolarcar.ca>
+ */
+
+#ifndef DeltaSongPlayer_H
```
Comment (bill-luu) [inline src/BusinessLayer/DeltaSongPlayer.h]: Update to pragma

File: src/BusinessLayer/LibMpgMediaPlayer.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#include "LibMpgMediaPlayer.h"
+#include <mpg123.h>
+#include <string>
+#include <cstring>
+#include <unistd.h>
```
Comment (bill-luu) [inline src/BusinessLayer/LibMpgMediaPlayer.cpp]: We might be able to remove a few of these unused includes

File: src/BusinessLayer/LibMpgMediaPlayer.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        return songplayer->currentSongTitle();
+    }
+    else if (key == QMediaMetaData::AlbumTitle)
+    {
+        return songplayer->currentSongAlbum();
+    }
+    else
+    {
+        return "Ben is Fat";
```
Comment (bill-luu) [inline src/BusinessLayer/LibMpgMediaPlayer.cpp]: lol I can't believe Ben didn't see this

File: src/BusinessLayer/LibMpgMediaPlayer.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+#include <QMediaPlayer>
+#include <QThread>
+#include "SongPlayer/I_MediaPlayer.h"
+#include "SongControl/SongControl.h"
+#include <mpg123.h>
+#include "DeltaSongPlayer.h"
```
Comment (bill-luu) [inline src/BusinessLayer/LibMpgMediaPlayer.h]: Again, there might be a few unused includes here

File: src/BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 9]
```diff
@@ -119,37 +119,37 @@ void SongPlayer::playPrevious()
...
+//        else
+//        {
+//            openFile();
+//        }
 
-        mediaPlayer_.play();
-    }
-    else if (mediaPlayer_.state() == QMediaPlayer::PlayingState)
+//        mediaPlayer_->play();
+//    }
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayer/SongPlayer.cpp]: Remove these comments

File: src/BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 10]
```diff
@@ -160,12 +160,13 @@ void SongPlayer::durationChanged(qint64 duration)
 void SongPlayer::positionChanged(qint64 position)
 {
     position_ = position;
+//    qDebug() << "position: " << position;
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayer/SongPlayer.cpp]: Drop this comment

File: src/BusinessLayer/SongPlayer/SongPlayer.h [hunk 11]
```diff
@@ -16,8 +14,16 @@
...
 #include <QProgressBar>
 #include <QScopedPointer>
-#include <QMediaMetaData>
 #include "../SongControl/SongControl.h"
+#include "../LibMpgMediaPlayer.h"
+
+#include <mpg123.h>
+#include <string>
+#include <cstring>
+#include <unistd.h>
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayer/SongPlayer.h]: Miight not need these?

File: src/BusinessLayer/threadclass.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ *  You should have received a copy of the GNU Affero General
+ *  Public License along with the Schulich Delta OnBoard Media
+ *  Control.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  For further contact, email <software@calgarysolarcar.ca>
+ */
+
+#ifndef I_THREADCLASS_H
+#define I_THREADCLASS_H
```
Comment (bill-luu) [inline src/BusinessLayer/threadclass.h]: Switch to pragma

**Review by bill-luu — CHANGES_REQUESTED** [review]

File: src/BusinessLayer/LibMpgMediaPlayer.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        return songPlayer_->currentSongTitle();
+    }
+    else if (key == QMediaMetaData::AlbumTitle)
+    {
+        return songPlayer_->currentSongAlbum();
+    }
+    else
+    {
+        return "Ben is Fat";
```
Comment (bill-luu) [inline src/BusinessLayer/LibMpgMediaPlayer.cpp]: Lets be nice :) 
`return QString()`

File: src/BusinessLayer/SongPlayer/SongPlayer.cpp [hunk 14]
```diff
@@ -193,49 +181,55 @@ void SongPlayer::toggleLoop()
...
+    cover_ = (ALBUM_FILE_PATH);
+    cover_.append(album_);
 
     QPixmap img(cover_);
 
     emit updateGUI(title_, artist_, img);
 }
 
-QColor SongPlayer::getColor(QImage img)
+QColor SongPlayer::getColor(QImage img, int thread_ID)
```
Comment (bill-luu) [inline src/BusinessLayer/SongPlayer/SongPlayer.cpp]: Keep strict camelCasing for argument names. `threadId`

**Review by bill-luu — CHANGES_REQUESTED** [review]
Looking good! Last few comments.

**Review by bill-luu — APPROVED** [review]

**Review by bill-luu — APPROVED** [review]

**Pull request merged**